enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -254,31 +253,39 @@ impl<'a> Handler<'a> {
     pub fn read(&mut self) -> u64 {
         self.trace("read", 3);
 
-        let fd = self.aex.gpr.rdi.raw();
-        let buf = self.aex.gpr.rsi.raw() as *mut u8;
-        let size = self.aex.gpr.rdx.raw();
-
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size) {
-            Err(errno) => return errno as u64,
-            Ok(map) => map,
+        // Allocate some unencrypted memory from the Block.
+        let cursor = self.block.cursor();
+        let slice: &mut [u8] = match unsafe { cursor.alloc(self.aex.gpr.rdx.into()) } {
+            Ok(slice) => slice,
+            Err(_) => return -libc::EMSGSIZE as u64,
         };
 
-        unsafe {
-            // Do the syscall; replace encrypted memory with unencrypted memory.
-            let ret = self.syscall(libc::SYS_read as u64, fd, map as _, size, 0, 0, 0);
+        // Do the syscall; replace encrypted memory with unencrypted memory.
+        let req = request!(libc::SYS_read => self.aex.gpr.rdi, slice, slice.len());
+        let res = unsafe { self.proxy(req) };
 
-            // Copy the unencrypted input into encrypted memory.
-            if ret <= ERRNO_BASE {
-                if ret > size {
+        match res {
+            Ok(res) => {
+                if usize::from(res[0]) > self.aex.gpr.rdx.into() {
```

Comment:
Is `res` always guaranteed to be non-empty? If not, index access into an empty container will result in a panic. Either way, I think it is almost always better to use `res.get(0).expect("some msg")` than `res[0]` since it aids in auditing the code.

Reply:
This is a good point; I'll update it. Thanks @kubkon !

Reply:
@npmccallum The above approach lets us avoid panics, but I think it doesn't work with the `Register` conversions so I'm having to do casts again if I use it - thoughts?

Reply:
@kubkon The `res` value is a `[Register<usize>; 2]` which conforms to the Linux syscall ABI. It will never be empty.

Reply:
@npmccallum Ah, OK! Thanks for the explanation!